### PR TITLE
Refactor getVisibleFunctions()

### PR DESCRIPTION
**Pull request by vasslitvinov**
This supports my work on point of instantiation.
The refactor reduces code duplication and splits
getVisibleFunctions() into manageable chunks.

Business logic remains the same except for a corner case
at the end of getVisibleFunctions(). There, we now check
the instantiation point upon rootModule->block whereas
before sometimes we did not. This change should not matter
because duplicate checking of a scope is a no-op. Also
there should be no instantiation point for rootModule->block,
to begin with.

Future work: getVisibleMethods() could also be switched
to using most of the same helper functions, even though
it is noticeably different due to foregoing privacy checks
and inUseChain.

Testing: standard and gasnet over multilocale tests.

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@lydia-duncan would you review?

File: compiler/resolution/visibleFunctions.cpp [hunk 1]
```diff
@@ -479,297 +479,238 @@ void getVisibleFunctions(const char*      name,
...
+static void getVisibleFunctions(const char*           name,
+                                CallExpr*             call,
+                                BlockStmt*            block,
+                                std::set<BlockStmt*>& visited,
+                                Vec<FnSymbol*>&       visibleFns,
+                                bool                  inUseChain)
+{
+  const bool firstVisit = (visited.find(block) == visited.end());
 
-      if (instantiationPt != NULL) {
```
Comment (lydia-duncan) [inline compiler/resolution/visibleFunctions.cpp:660]: The scope of this removed `if` is different than the scope of the same check that was removed on line 767.  I think that means that if the block being visited for the first time is the rootModule's block, we previously didn't check the instantiation point but now will do so.  I don't know what the implications of that are (and maybe the new behavior isn't a bad thing), but I suspect it was an edge case we didn't have a test for so would like to make sure you're aware of it since the PR message claims the business logic is unchanged.

**Comment by lydia-duncan** [comment]
Other than the question I had, I think this is reasonable, thanks!
